File: backend/app/api/data/dividends.py
"""Dividend / long-run compounding endpoints."""
from __future__ import annotations

from fastapi import APIRouter, Query

from app.data.market import dividends
from app.data.market import dividend_detail
from app.data.market import dividend_royalty
from app.data.market import dividend_etf
from app.data.market import crisis_survivors

router = APIRouter()


@router.get("/dividends")
def dividends_endpoint():
    """배당·실적 — 고배당 랭킹 + 영업이익 YoY 실적개선 랭킹."""
    return dividends.board()


@router.get("/dividend-universe")
def dividend_universe_endpoint():
    """종목 단위 배당 계산기 — 검색 가능한 전 종목 + 배당수익률·추정 주당배당금(DPS)."""
    return dividends.stock_universe()


@router.get("/dividend-detail")
def dividend_detail_endpoint(ticker: str = Query(..., description="single ticker")):
    """종목 단위 배당 심층 분석 — 배당률 + 투자전 체크리스트(매출·순이익·영업현금흐름·
    배당연수·배당성장률) + 3대 위기(2000·2008·2020) 배당 내역."""
    return dividend_detail.detail(ticker)


@router.get("/dividend-royalty")
def dividend_royalty_endpoint(invest: float = Query(default=0, ge=0)):
    """배당 성장주 레퍼런스 — 배당왕(50년+)·배당귀족(25년+ & S&P500)·월배당 포트.

    invest>0 이면 월배당 동일가중 포트폴리오의 월 배당(세전/세후) 추정도 함께 반환.
    """
    out = dividend_royalty.board()
    if invest > 0:
        out["portfolio"] = dividend_royalty.monthly_portfolio(invest)
    return out


@router.get("/crisis-survivors")
def crisis_survivors_endpoint():
    """3대 위기(2000·2008·2020)를 이겨내고 우상향한 배당주 — 장기 주가 궤적 +
    위기별 낙폭 + 배당 방어(배당왕/귀족). 첫 호출은 주가 취합으로 느릴 수 있음(하루 캐시)."""
    return crisis_survivors.board()


@router.get("/dividend-etf")
def dividend_etf_endpoint():
    """배당·인덱스 ETF 레퍼런스 — 배당성장(VIG·SCHD·DGRO 등)·고배당·커버드콜·S&P500."""
    return dividend_etf.board()


@router.get("/sp-dca")
def sp_dca_endpoint(
    monthly: float = Query(default=500000, ge=0, description="월 적립액(원 또는 통화단위)"),
    years: float = Query(default=20, gt=0, le=60),
    annual_return: float = Query(default=0.10, ge=-0.5, le=0.5),
):
    """S&P500 적립형 — 매월 적립 시 미래가치·예상 배당 추정(복리)."""
    return dividend_etf.sp_dca(monthly, years, annual_return)

File: frontend/src/lib/api/dividends.ts
// 배당 — 배당주 보드·심층분석·배당왕/귀족·위기생존주·ETF·S&P 적립

import { request } from "./client";

export interface DividendRow {
  ticker: string;
  name: string | null;
  sector: string | null;
  close: number | null;
  div_yield: number;
  per: number | null;
  roe: number | null;
}

export interface EarningRow {
  ticker: string;
  name: string | null;
  sector: string | null;
  close: number | null;
  period: string;
  op_yoy: number;
  op_margin: number | null;
  op_profit: number | null;
}

export interface DividendsBoard {
  generated_at: string;
  dividends: DividendRow[];
  earnings: EarningRow[];
  note: string;
}

export interface DividendStock {
  ticker: string;
  name: string;
  sector: string | null;
  close: number;
  div_yield: number | null;
  dps: number | null;
}

export interface DividendUniverse {
  generated_at: string;
  count: number;
  stocks: DividendStock[];
  note: string;
}

// ── 종목 단위 배당 심층 분석 ──────────────────────────────────────────────
export interface DDMetric {
  series: { year: number; value: number; estimate: boolean }[];
  latest: { year: number; value: number } | null;
  trend: "증가" | "감소" | "정체" | null;
  unit: string;
  why: string;
  available?: boolean;
  note?: string;
}

export interface DDCrisisRow { year: number; dps: number | null; verdict: "증가" | "유지" | "삭감" | "중단" | null; }

export interface DDCrisis {
  key: string;
  label: string;
  rows: DDCrisisRow[];
  summary: string;
  min?: number | null;
  max?: number | null;
}

export interface DividendDetail {
  ticker: string;
  name: string | null;
  sector: string | null;
  market?: "KR" | "US";
  currency?: "KRW" | "USD";
  royalty?: { tier: string; tier_label: string; years: number | null } | null;
  close: number | null;
  generated_at: string;
  dividend: { dps: number | null; dps_estimated: boolean; div_yield: number | null; formula: string };
  checklist: {
    revenue: DDMetric;
    net_income: DDMetric;
    op_cash_flow: DDMetric;
    div_years: { value: number; window: [number, number] | null; why: string };
    div_growth: { cagr: number | null; series: { year: number; dps: number }[]; window: [number, number] | null; why: string };
    roe: DDMetric;
  } | null;
  crises: { available: boolean; name: string | null; notes: string | null; sources: string[]; crises: DDCrisis[] } | null;
  note: string;
}

// ── 배당왕·귀족·월배당 ────────────────────────────────────────────────────
export interface RoyaltyRow { ticker: string; name: string; sector?: string; type?: string; years?: number | null; yield?: number | null; freq?: string; }

export interface RoyaltyGroup { count: number; criteria: string; avg_yield: number | null; rows: RoyaltyRow[]; }

export interface MonthlyPortfolio {
  invest: number; blended_yield: number; annual_gross: number; annual_net: number;
  monthly_gross: number; monthly_net: number; n_holdings: number; note: string;
}

export interface DividendRoyalty {
  as_of: string;
  kings: RoyaltyGroup;
  aristocrats: RoyaltyGroup;
  monthly: RoyaltyGroup;
  portfolio?: MonthlyPortfolio;
  note: string;
}

// ── 위기를 이겨낸 우상향 배당주 ───────────────────────────────────────────
export interface SurvivorCrisis { key: string; label: string; drawdown: number | null; dividend: string; }

export interface SurvivorRow {
  ticker: string; name: string; sector: string | null; tier_label: string | null; years: number | null;
  multiple: number | null; cagr: number | null;
  index: { date: string; v: number }[];
  crises: SurvivorCrisis[];
}

export interface CrisisSurvivors {
  generated_at: string; start: string; benchmark: SurvivorRow | null;
  survivors: SurvivorRow[]; crises: { key: string; label: string }[]; note: string;
}

// ── 배당 ETF + S&P 적립 ───────────────────────────────────────────────────
export interface EtfRow {
  ticker: string; name: string; category: string; yield: number | null;
  div_cagr_5y: number | null; expense: number | null; inception: number | null;
  freq: string; strategy: string;
}

export interface EtfGroup { category: string; count: number; avg_yield: number | null; rows: EtfRow[]; }

export interface DividendEtfBoard { as_of: string; groups: EtfGroup[]; count: number; note: string; }

export interface SpDca {
  monthly: number; years: number; annual_return_pct: number; principal: number;
  future_value: number; gain: number; est_annual_dividend: number; est_monthly_dividend: number; note: string;
}

export const dividendsApi = {
  dividends: () => request<DividendsBoard>(`/api/data/dividends`),
  dividendUniverse: () => request<DividendUniverse>(`/api/data/dividend-universe`),
  dividendDetail: (ticker: string) => request<DividendDetail>(`/api/data/dividend-detail?ticker=${ticker}`),
  dividendRoyalty: (invest = 0) => request<DividendRoyalty>(`/api/data/dividend-royalty${invest > 0 ? `?invest=${invest}` : ""}`),
  crisisSurvivors: () => request<CrisisSurvivors>(`/api/data/crisis-survivors`),
  dividendEtf: () => request<DividendEtfBoard>(`/api/data/dividend-etf`),
  spDca: (monthly: number, years: number, annualReturn: number) =>
    request<SpDca>(`/api/data/sp-dca?monthly=${monthly}&years=${years}&annual_return=${annualReturn}`),
};
